crates/bevy_render/src/render_resource/storage_buffer.rs

```diff
@@ -0,0 +1,138 @@
+    }
+
+    #[inline]
+    pub fn set_body(&mut self, body: T) {
+        self.body = body;
+    }
+
+    fn reserve_buffer(&mut self, device: &RenderDevice) -> bool {
+        let size = self.size();
+        if self.storage_buffer.is_none() || size > self.scratch.len() {
+            self.scratch.resize(size, 0);
+            self.storage_buffer = Some(device.create_buffer(&BufferDescriptor {
+                label: None,
+                size: size as wgpu::BufferAddress,
+                usage: BufferUsages::COPY_DST | BufferUsages::STORAGE,
+                mapped_at_creation: false,
+            }));
+            true
+        } else {
+            false
+        }
+    }
+
+    fn size(&self) -> usize {
+        let mut size = 0;
+        size += Self::BODY_SIZE;
+        if Self::ITEM_SIZE > 0 {
+            if size > 0 {
+                // Pad according to the array item type's alignment
+                size = (size + <U as AsStd430>::Output::ALIGNMENT - 1)
+                    & !(<U as AsStd430>::Output::ALIGNMENT - 1);
+            }
+            // Variable size arrays must have at least 1 element
+            size += Self::ITEM_SIZE * self.values.len().max(1);
+        }
+        size
+    }
+
+    pub fn write_buffer(&mut self, device: &RenderDevice, queue: &RenderQueue) {
+        self.reserve_buffer(device);
+        if let Some(storage_buffer) = &self.storage_buffer {
+            let range = 0..self.size();
+            let mut writer = std430::Writer::new(&mut self.scratch[range.clone()]);
+            let mut offset = 0;
+            // First write the struct body if there is one
+            if Self::BODY_SIZE > 0 {
+                if let Ok(new_offset) = writer.write(&self.body).map_err(|e| warn!("{:?}", e)) {
+                    offset = new_offset;
+                }
+            }
+            if Self::ITEM_SIZE > 0 {
+                if self.values.is_empty() {
+                    // Zero-out the padding and dummy array item in the case of the array being empty
+                    for i in offset..self.size() {
+                        self.scratch[i] = 0;
+                    }
+                } else {
+                    // Then write the array. Note that padding bytes may be added between the body
+                    // and the array in order to align the array to the alignment requirements of its
+                    // items
+                    writer
+                        .write(self.values.as_slice())
+                        .map_err(|e| warn!("{:?}", e))
+                        .ok();
+                }
+            }
+            queue.write_buffer(storage_buffer, 0, &self.scratch[range]);
+        }
+    }
+}
+
+impl<T: AsStd430, U: AsStd430> Deref for StorageBuffer<T, U> {
+    type Target = Vec<U>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.values
+    }
+}
+
+impl<T: AsStd430, U: AsStd430> DerefMut for StorageBuffer<T, U> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
```

Comment:
I don't think this should use the Deref / DerefMut approach like UniformVec does (I'm not even fully sold yet on that approach for UniformVec). Things that impl Deref/DerefMut should only deref into "equivalent types" (ex: smart pointers deref to their actual types, wrapper tuple structs deref into their wrapped type). UniformVec is easier to defend because it really is a "vec-like" api (at least for now). But StorageBuffer is a "struct like" api that happens to have an optional vector as a "field". Imo that is very confusing.

Reply:
You mean that its deref behaviour is confusing and you would prefer to have explicit methods for interacting with the Vec?

Reply:
Yeah. I think the vec should be exposed via accessor methods.

Reply:
Done via `values() -> &[T]` and `values_mut() -> &mut [T]`